### PR TITLE
Added test for pduToSample and corrected int to uint for metric value

### LINKED ISSUE #191: Negative value for ifHCOutUcastPkts
Hi,

I have the snmp_exporter running against a Juniper MX (Junos 17.1) with the default snmp.yaml and I'm getting a negative value for ifHCOutUcastPkts:

# HELP ifHCOutUcastPkts 
# TYPE ifHCOutUcastPkts counter
ifHCOutUcastPkts{ifDescr="ae0"} -1.241915e+06

I verified on the router that the value is indeed not negative:

teichhorn@DE-ES-001-03-10-01-re0> show snmp mib get 1.3.6.1.2.1.31.1.1.1.11.629
ifHCOutUcastPkts.629 = 18446744073708309388

I am not sure if this is a problem handling 64bit values, 
or howto debug this further - running the exporter with log level debug doesn't give any hints to what is happening behind the curtains.

If you need any help from my side, please don't hesitate to ask.

Thanks,
Tom

### PR DESCRIPTION
**Pull request by Conorbro**
Fixes #191

### CODE REVIEW

File: collector_test.go [hunk 1]
```diff
@@ -9,6 +9,72 @@ import (
...
+				Help: "Help string",
+			},
+			oidToPdu:       make(map[string]gosnmp.SnmpPDU),
+			expectedResult: "Desc{fqName: \"test_metric\", help: \"Help string\", constLabels: {}, variableLabels: [test_metric]}",
+		},
+	}
+
+	for _, c := range cases {
+		res := pduToSample(c.indexOids, c.pdu, c.metric, c.oidToPdu)
+		if res.Desc().String() != c.expectedResult {
```
Comment (brian-brazil) [inline collector_test.go]: You'd probably be better off using `Write` rather than `Desc` so the values can be checked.

File: collector_test.go [hunk 2]
```diff
@@ -9,6 +9,72 @@ import (
...
+		oidToPdu       map[string]gosnmp.SnmpPDU
+		expectedResult string
+	}{
+		{
+			pdu: &gosnmp.SnmpPDU{
+				Name:  "1.3.6.1.2.1.31.1.1.1.11.629",
+				Type:  0x02,
+				Value: 2,
+			},
+			indexOids: []int{1},
```
Comment (brian-brazil) [inline collector_test.go]: This should be empty, as there's no index in the config.Metric

File: collector_test.go [hunk 3]
```diff
@@ -9,6 +9,72 @@ import (
...
+	cases := []struct {
+		pdu            *gosnmp.SnmpPDU
+		indexOids      []int
+		metric         *config.Metric
+		oidToPdu       map[string]gosnmp.SnmpPDU
+		expectedResult string
+	}{
+		{
+			pdu: &gosnmp.SnmpPDU{
+				Name:  "1.3.6.1.2.1.31.1.1.1.11.629",
```
Comment (brian-brazil) [inline collector_test.go]: Use dummy oids for testing, otherwise you have to make sure the data makes sense for this oid to have

**Comment by Conorbro** [comment]
@brian-brazil

File: collector_test.go [hunk 4]
```diff
@@ -6,9 +6,82 @@ import (
...
+				Oid:  "1.1.1.1.1",
+				Help: "Help string",
+			},
+			oidToPdu: make(map[string]gosnmp.SnmpPDU),
+		},
+	}
+
+	for _, c := range cases {
+		m := pduToSample(c.indexOids, c.pdu, c.metric, c.oidToPdu)
+		err := m.Write(&io_prometheus_client.Metric{})
```
Comment (brian-brazil) [inline collector_test.go]: You should check what's put in the metric.

**Comment by Conorbro** [comment]
@brian-brazil

File: collector_test.go [hunk 5]
```diff
@@ -6,9 +6,90 @@ import (
...
+				Value: -2,
+			},
+			indexOids: []int{},
+			metric: &config.Metric{
+				Name: "test_metric",
+				Oid:  "1.1.1.1.1",
+				Help: "Help string",
+			},
+			oidToPdu:       make(map[string]gosnmp.SnmpPDU),
+			expectedMetric: "label:<name:\"test_metric\" value:\"-2\" > gauge:<value:1 > ",
```
Comment (brian-brazil) [inline collector_test.go]: You can use backticks to have a raw value, so you don't need the \ for escaping

File: collector_test.go [hunk 6]
```diff
@@ -6,9 +6,90 @@ import (
...
+		}
+		if metric.String() != c.expectedMetric {
+			t.Fatalf("Unexpected metric: got %v, want %v", metric.String(), c.expectedMetric)
+		}
+	}
+}
+
+func TestGetPduValue(t *testing.T) {
+	pdu := &gosnmp.SnmpPDU{
+		Value: -2,
```
Comment (brian-brazil) [inline collector_test.go]: Signed integers are supported by SNMP, it's the Counter64 which are the uint64s that we need to be careful with.

File: collector_test.go [hunk 7]
```diff
@@ -6,9 +6,90 @@ import (
...
+				Type: "counter",
+				Help: "Help string",
+			},
+			oidToPdu:       make(map[string]gosnmp.SnmpPDU),
+			expectedMetric: "counter:<value:2 > ",
+		},
+		{
+			pdu: &gosnmp.SnmpPDU{
+				Name:  "1.1.1.1.1",
+				Type:  0x02,
```
Comment (brian-brazil) [inline collector_test.go]: Use the names here for clarity: https://github.com/soniah/gosnmp/blob/adbb107aeaa6c23e71056bfd7a095874a07c7d04/gosnmp.go#L134

**Comment by Conorbro** [comment]
@brian-brazil

File: collector_test.go (hunk 6 above)
Comment (brian-brazil) [inline collector_test.go]: This test doesn't make sense, -2 isn't a valid value for a Counter64. Try 2^63.

File: collector_test.go [hunk 8]
```diff
@@ -6,9 +6,91 @@ import (
...
+			},
+			indexOids: []int{},
+			metric: &config.Metric{
+				Name: "test_metric",
+				Oid:  "1.1.1.1.1",
+				Type: "counter",
+				Help: "Help string",
+			},
+			oidToPdu:       make(map[string]gosnmp.SnmpPDU),
+			expectedMetric: "counter:<value:2 > ",
```
Comment (brian-brazil) [inline collector_test.go]: Should the help string not get propagated?

Comment (Conorbro) [inline collector_test.go]: The `io_prometheus_client.Metric` type doesn't actually have a field for help strings.

Comment (brian-brazil) [inline collector_test.go]: Hmm, we'll need to check the output of Desc as well then.

File: collector_test.go [hunk 9]
```diff
@@ -1,14 +1,97 @@
...
+		}
+		if metric.String() != c.expectedMetric {
+			t.Fatalf("Unexpected metric: got %v, want %v", metric.String(), c.expectedMetric)
+		}
+	}
+}
+
+func TestGetPduValue(t *testing.T) {
+	pdu := &gosnmp.SnmpPDU{
+		Value: uint64(math.Pow(2, 63)),
```
Comment (brian-brazil) [inline collector_test.go]: `uint64(1 << 63)` would be better, as it avoids floating point.

**Comment by Conorbro** [comment]
@brian-brazil

**Comment by brian-brazil** [comment]
Thanks!

**Pull request merged**